### PR TITLE
[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* __->__ #169454

While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.

### CODE REVIEW

File: torch/distributed/_pycute/layout.py [hunk 1]
```diff
@@ -167,12 +167,13 @@ def coalesce(layout: Layout, profile: LayoutProfile = None) -> Layout:
             )
         )
 
+    flattened_stride = flatten(layout.stride)
     result_shape = [1]
-    result_stride = [0]
+    result_stride = [flattened_stride[-1]]
```
Comment (lw) [inline torch/distributed/_pycute/layout.py]: You said

> When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong

I disagree with that claim. One of the "purposes" of coalescing is also to normalize a layout and produce a canonical representation. This is key because we (might) use the layout as a cache key hence we need to be able to deduplicate layouts if they are effectively equivalent.

If here we take `stride[-1]` instead of 0, we might break this property. For example, layouts (1, 1):(42, 3) and (1, 1, 1):(17, 4, 13) are, in fact, the same layout, but after this PR they coalesce to two _different_ layouts (i.e., (1,):(3,) and (1,):(13,) respectively). This is not acceptable. (In fact, we might want to add a test case to check this property)

Note, also, that if the size of a layout is (1,), then the stride is _irrelevant_, and any value is acceptable. Which I believe is why the original code was using 0 as a "default" stride.

Now, I wonder why this "fix" was needed in the first place. Does another part of the code fail when given a stride of 0? If so, we likely need to fix that piece of code instead...

File: torch/distributed/device_mesh.py [hunk 2]
```diff
@@ -675,21 +675,17 @@ def _create_sub_mesh(
...
             for name in submesh_dim_names:
                 if name in not_none(self._mesh_dim_names):
-                    slice_dim_group_name.append(
-                        self._dim_group_names[  # type: ignore[has-type]
-                            not_none(self._mesh_dim_names).index(name)
-                        ]
-                    )
+                    idx = not_none(self._mesh_dim_names).index(name)
+                    if idx < len(self._dim_group_names):
+                        slice_dim_group_name.append(self._dim_group_names[idx])
```
Comment (lw) [inline torch/distributed/device_mesh.py]: Can you explain again in which cases this is necessary?

In general, I'm very surprised to find out that `self._mesh_dim_names` and `self._dim_group_names` could have a different length, and if this is indeed allowed then it could be the source of all sorts of pitfalls.

I can see that in general we have two cases:
- either the local rank belongs to a given mesh, in which case it necessarily belongs to all the PGs of the mesh, thus `len(_dim_group_names) == len(_mesh_dim_names)`;
- or the local rank does _not_ belong to a mesh, but in that case it doesn't belong to _any_ of the PGs, thus `len(_dim_group_names) == 0`.
Is my understanding correct?

If so, I'd prefer if we rewrite the code to make it clear that we're simply checking in which of the two situations above we fall. For example, we can simply gate the entire for loop behind a check that `_dim_group_names` is not empty:
```py
if len(self._dim_group_names) > 0:
    assert len(self._dim_group_names) == len(self._mesh_dim_names)
    for name in submesh_dim_names:
        ...
```

**Review by lw — CHANGES_REQUESTED** [review]
I'm a bit suspicious of at least one of the fixes and would like to understand more before approving.

File: torch/distributed/_pycute/layout.py (hunk 1 above)
Comment (cspades) [reply torch/distributed/_pycute/layout.py]: @lw The other part of the code that fails is this: https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L847-L851
```
                if stride < pre_stride:
                    raise KeyError(
                        f"Invalid mesh_dim_names {mesh_dim_names} specified. "
                        "Mesh dim indices should be in ascending order."
                    )
```
If we need 0 as a natural equivalence class for stride when coalescing shapes, I don't think we can really depend on stride for any kind of logic like this right? (I'm not even sure what this error even accomplishes. What exactly is stride in the context of mesh dimensions, and why do we need those dimensions sorted by stride? Couldn't find super clear documentation on this.)

My original issue: https://github.com/pytorch/pytorch/issues/169381#issue-3686620194

Comment (lw) [reply torch/distributed/_pycute/layout.py]: I'll take a look at the original issue tomorrow, but I suspect it can be fixed more robustly by coalescing `slices_sizes` and `slices_strides` here: https://github.com/pytorch/pytorch/blob/7348cb355ff0a6f79cd4871215aea72185748734/torch/distributed/device_mesh.py#L831-L832

Since the error you encountered would only be raised if a 0-stride dimension occurs in the middle of a layout, but this is precisely what coalesce helps prevent.

Note also that the whole check could also just be removed (there's a TODO which suggests doing just that), we simply haven't done it yet because we don't want to expand the API surface without a use-case.

Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: @lw I agree with the example you mentioned here. but the previous implementation also does not make sense, if we coalesce a layout (1:1) the result will be (1:0) so the stride will be 0 this is also wrong right?

And for the case you mentioned, layouts (1, 1):(42, 3) and (1, 1, 1):(17, 4, 13), for the current code logic, they will all be (1:0) after coalesce which is also weird, I think we need to define this correctly for the size of 1. (I do agree stride does not matter in this case since the size is only 1)

I think what you propose would work as well.

File: torch/distributed/device_mesh.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/device_mesh.py]: The reason is that in this case, world_size is 8 but the mesh size is only 1, so only rank 0 will create PG and all other ranks will have None returned. So the _dim_group_names will be empty list on non-rank 0. This is a corner case when the dim of mesh is smaller than world size.

After thinking about it, I think your proposal would work, and we should add this explicit check as well

File: torch/distributed/_pycute/layout.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: @lw I played with PyTorch, when I do `torch.ones(1,1,1)` I got strides to be (1,1,1) so pytorch is using stride to be 1 for the size of 1.

File: torch/distributed/device_mesh.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/device_mesh.py]: Well, we will still need this inside else for flattened mesh names as well... 

NVM I think even if it is a submesh flattened, the rank is still somehow in the mesh.. we only need that one check.

File: torch/distributed/_pycute/layout.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: > Note also that the whole check could also just be removed (there's a TODO which suggests doing just that), we simply haven't done it yet because we don't want to expand the API surface without a use-case

The reason we don't remove it is because it is an undefined behavior which has not been supported previously.

Comment (lw) [reply torch/distributed/_pycute/layout.py]: > if we coalesce a layout (1:1) the result will be (1:0) so the stride will be 0 this is also wrong right?

In CuTe's "definition" of a Layout, using a stride of 0 is **correct**. I think it all stems from the fact that a _size_ of 1 is "weird", for two reasons: it is compatible with _any_ stride (i.e., layouts `1:1` and `1:42` are _identical_), and thus it is the _only_ size which is compatible with a stride of 0.

Another way of saying it is that, when we "normalize" a layout, what we would like to do is take all the dimensions that have size 1 and set their stride to a special wildcard `*`, which means that in fact we don't care about the value of that stride. CuTe's convention is to use 0 instead of that wildcard.

That said, if we want to change the convention we can, and we can use a stride of 1 instead of that wildcard if you prefer. However, doing so should be done carefully since we'd need to check the entirety of pycute to see if other places depend on this convention. In general, I'd advise _against_ changing that convention as it causes our copy of pycute to diverge from the upstream one.

Before embarking in all of this, could you attempt another potential fix? Could you try adding the following lines just _before_ the problematic `for stride in reversed(sliced_strides):` loop?
```py
sliced_layout = torch.distributed._pycute.coalesce(torch.distributed._pycute.Layout(sliced_sizes, sliced_strides))
sliced_sizes, sliced_strides = sliced_layout.shape, sliced_layout.strides
```

Comment (wconstab) [reply torch/distributed/_pycute/layout.py]: @fduwjj did you add this test case? i think it sounds like a good idea

> If here we take stride[-1] instead of 0, we might break this property. For example, layouts (1, 1):(42, 3) and (1, 1, 1):(17, 4, 13) are, in fact, the same layout, but after this PR they coalesce to two different layouts (i.e., (1,):(3,) and (1,):(13,) respectively). This is not acceptable. (In fact, we might want to add a test case to check this property)

Also, if the only problem that is caused by a 0 stride was tripping this assertion, can't we just workaround to special case stride=0 when implementing the assertion?

Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: @lw I don't think this will help because if the size is 1, we always skip during coalesce, so if the layout is ((1,1,1), (1,0,1)) the result will be the same

File: test/distributed/_pycute/test_coalesce.py [hunk 3]
```diff
@@ -47,65 +47,82 @@
...
+        expected_coalesced_layout = Layout(1, 0)
+        self.helper_test_coalesce(layout, expected_coalesced_layout)
+
+        layout = Layout((1, 1), (42, 3))
+        expected_coalesced_layout = Layout(1, 0)
+        self.helper_test_coalesce(layout, expected_coalesced_layout)
+
+        layout = Layout((1, 1, 1), (17, 4, 3))
+        expected_coalesced_layout = Layout(1, 0)
+        self.helper_test_coalesce(layout, expected_coalesced_layout)
```
Comment (fduwjj) [inline test/distributed/_pycute/test_coalesce.py:77]: @wconstab the tests @lw asked is added here.

File: torch/distributed/_pycute/layout.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: ok I end up changing it inside _mesh_layout so that we don't allow 0 stride for device mesh

Comment (lw) [reply torch/distributed/_pycute/layout.py]: I have another idea for solving this problem, which actually matches _even more closely_ what PyTorch does on regular Tensors. The idea behind removing dimensions with size == 1 is the same as calling `unsqueeze` on a tensor: these dimensions have no effect on the overall numel of the tensor, thus can be removed. If we have a Tensor with only size-1 dimensions and we call unsqueeze on all of them, we end up with a zero-dimensional tensor! Perhaps we should make pycute does the same!

Concretely: calling coalesce on a layout (1, 1, 1):(42, 37, 1) now return 1:0, instead it could just return ():(). _In principle_ the effect should be the same, but we might need to make sure that the rest of our code is able to handle empty (0-dimensional) layouts. WDYT?

File: torch/distributed/device_mesh.py [hunk 4]
```diff
@@ -676,23 +676,30 @@ def _create_sub_mesh(
...
-                    # Since we will deprecate the slicing of flattened dim_name from root mesh soon,
-                    # we don't want to optimize the code furthermore.
-                    flatten_mesh = self._flatten_mapping[name]
-                    slice_dim_group_name.append(
-                        flatten_mesh._dim_group_names[  # type: ignore[has-type]
-                            not_none(flatten_mesh._mesh_dim_names).index(name)
-                        ]
-                    )
+            if len(self._dim_group_names) > 0:
+                assert len(self._dim_group_names) == len(
```
Comment (lw) [inline torch/distributed/device_mesh.py:681]: Would it cause any problems to move this assert just before the `if`? This way it would also cover the case of both lists being empty.

File: torch/distributed/_mesh_layout.py [hunk 5]
```diff
@@ -104,7 +104,15 @@ def coalesce(self) -> "_MeshLayout":
...
-        return _MeshLayout(layout.shape, layout.stride)
+        # The original PuCute coalesce() will use stride=0 for size=1 for all dimension.
+        # We don't want to do that in device mesh, we will reset them to be 1 to be same as PyTorch.
+        if is_int(layout.stride) and layout.stride == 0:
+            return _MeshLayout(layout.shape, 1)
+        elif is_tuple(layout.stride) and any(s == 0 for s in layout.stride):
+            non_zero_strides = tuple(s if s != 0 else 1 for s in layout.stride)
+            return _MeshLayout(layout.shape, non_zero_strides)
+        else:
+            return _MeshLayout(layout.shape, layout.stride)
```
Comment (lw) [inline torch/distributed/_mesh_layout.py:115]: If we end up going with the idea of just completely pruning all size-1 dims from the layout (even at the cost of ending up with an empty 0-dim layout), could we turn this code in an assert to check that invariant?

File: torch/distributed/device_mesh.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/device_mesh.py:681]: that is not correct and this exactly what happens in this corner case, because only rank0 have `_dim_group_names`, all other ranks have empty `_dim_group_names` but still having `_mesh_dim_names`. So the reason is that mesh size == 1 while world_size == 4. So the first if just check whether one rank is within the mesh or not. Because the code we wrote now it is not both empty because we have device mesh object on all ranks.

File: torch/distributed/_mesh_layout.py (hunk 5 above)
Comment (fduwjj) [reply torch/distributed/_mesh_layout.py:115]: I think it is not just about `pruning all size-1 dims from the layout`. It is also about when we do flatten among certain dims and the rest of dims are still there. So you can not remove all these size-1 dims because they are actual dims of the mesh.

File: torch/distributed/_pycute/layout.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_pycute/layout.py]: > calling coalesce on a layout (1, 1, 1):(42, 37, 1) now return 1:0, instead it could just return ():()

This is a bad idea, because we still support flatten and slicing flattened dim from root for BC reason. In a world without BC then yes, that is a good idea. Why? Because first, you still need to change the coalesce inside mesh_layout so it does not make changes any smaller. Second, we need to add extra logic to handle that. I don't think we have empty size/stride for mesh_layout today, it is a way bigger change.

**Review by lw — APPROVED** [review]
I'm ok to go with stride=1 instead of stride=0 as a conventional value for size=1 dims. It seems harmless enough and preserves the "normalization" property of coalescing.
